_inc/lib/core-api/wpcom-endpoints/memberships.php

```diff
@@ -0,0 +1,170 @@
+
+	/**
+	 * WPCOM_REST_API_V2_Endpoint_Memberships constructor.
+	 */
+	public function __construct() {
+		$this->namespace                       = 'wpcom/v2';
+		$this->rest_base                       = 'memberships';
+		$this->wpcom_is_wpcom_only_endpoint    = true;
+		$this->wpcom_is_site_specific_endpoint = true;
+		add_action( 'rest_api_init', array( $this, 'register_routes' ) );
+	}
+
+	/**
+	 * Called automatically on `rest_api_init()`.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/status',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_status' ),
+					'permission_callback' => array( $this, 'get_status_permission_check' ),
+				),
+			)
+		);
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base . '/product',
+			array(
+				array(
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => array( $this, 'create_product' ),
+					'permission_callback' => array( $this, 'get_status_permission_check' ),
+					'args'                => array(
+						'title'    => array(
+							'type'     => 'string',
+							'required' => true,
+						),
+						'price'    => array(
+							'type'     => 'float',
+							'required' => true,
+						),
+						'currency' => array(
+							'type'     => 'string',
+							'required' => true,
+						),
+						'interval' => array(
+							'type'     => 'string',
+							'required' => true,
+						),
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Ensure the user has proper permissions
+	 *
+	 * @return boolean
+	 */
+	public function get_status_permission_check() {
+		return current_user_can( 'edit_posts' );
+	}
+
+	/**
+	 * Do create a product based on data, or pass request to wpcom.
+	 *
+	 * @param object $request - request passed from WP.
+	 *
+	 * @return array|WP_Error
+	 */
+	public function create_product( $request ) {
+		if ( ( defined( 'IS_WPCOM' ) && IS_WPCOM ) ) {
+			require_lib( 'memberships' );
+			$connected_destination_account_id = Jetpack_Memberships::get_connected_account_id();
+			if ( ! $connected_destination_account_id ) {
+				return new WP_Error( 'no-destination-account', __( 'Please set up Stripe account for this site first', 'jetpack' ) );
+			}
+			$product = Memberships_Product::create(
+				get_current_blog_id(),
+				array(
+					'title'                            => $request['title'],
+					'price'                            => $request['price'],
+					'currency'                         => $request['currency'],
+					'interval'                         => $request['interval'],
+					'connected_destination_account_id' => $connected_destination_account_id,
+				)
+			);
+			return $product->to_array();
+		} else {
+			$blog_id  = Jetpack_Options::get_option( 'id' );
+			$response = Jetpack_Client::wpcom_json_api_request_as_user(
```

Comment:
This causes issues when someone who did not connect their account to WordPress.com tries to add a button. They receive the generic error below because their account is not connected to WordPress.com (even if the site is actually connected).

Reply:
Good catch!
FWIW I tried using `wpcom_json_api_request_as_blog` and it is broken - does not work with v2 endpoints. But for now, I'll think of how we can limit access to this or provide a more meaningful error.
The multiple user scenario on JP needs some case studies and there competing theories on how this should work.

Reply:
Added better feedback.
https://github.com/Automattic/jetpack/pull/9802/commits/856fd58a6c0655173e86abc1033eed8eea5682f8